config/fcos/v1_5_exp/schema.go

```diff
@@ -38,3 +39,5 @@ type BootDeviceLuks struct {
 type BootDeviceMirror struct {
 	Devices []string `yaml:"devices"`
 }
+
+type Extensions []string
```

Comment:
Are we sure we won't have additional options related to extensions in the future?  This schema doesn't allow for any.

The usual way we do this is:
```yaml
extensions:
  - name: foo
  - name: bar
```

Reply:
We are basically adding something like the RHCOS extensions to butane: https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfiguration.md#rhcos-extensions .  But I am not sure how this will evolve. This is still being discussed but for the first pass AFAIK we just care about a list of packages.

Reply:
Once this spec version has been stabilized, we lose the ability to add metadata in this way without making a new top-level field.  We could merge this as-is and revisit before stabilization, but if we do, we should make sure to leave clear signposts in the issue tracker so we don't forget.  Or we could go for the extensible approach now.  It might be worth raising with the team.

Reply:
@cgwalters @jlebon any preference on how to handle this?

Reply:
Agree with the concern, but not sure what to recommend since it's hard to predict what we'll need to generalize (container-based extensions? more treefile support? repo-pinned extensions?).

So... I'd say let's roll with this for now and revisit before stabilization. In the worst case, we should be able to change the spec in future versions, right?

Reply:
Discussed this more with @jlebon OOB.  If we wanted to add extensibility in future versions, and didn't want to have a major spec bump (i.e. one that requires config changes when bumping the version number), we'd need to deprecate the `extensions` field and add e.g. `extensions2`.  We've done that before but it's a bit awkward.

I think the maximum possible extensibility is:

```yaml
extensions:
  packages:
    - name: foo
    - name: bar
```

which lets us add system-wide extension options, per-package ones, and new extension types:

```yaml
extensions:
  global-option: z
  packages:
    - name: foo
      package-option: z
    - name: bar
  containers:
    - name: baz
```

Global options don't play well with config merging, since a child config that adds a package would be affected by global parameters it isn't aware of.  New extension types could be implemented as a package field:

```yaml
extensions:
  - name: foo
    package-option: z
  - name: bar
  - name: baz
    type: container
```

which is consistent with the design elsewhere in the schema.

I think that's a reasonable middle ground.  Let's not block the current PR on this, but as a followup I think we should switch from an array of strings to an array of structs.

Reply:
This is very related to what happened to `packages` in 
https://github.com/coreos/rpm-ostree/blob/main/docs/treefile.md

We ended up adding `repo-packages` which allows to pin to packages from a specific repo.

--

The idea of `type: container` is a really interesting idea.  Definitely touches on https://www.freedesktop.org/software/systemd/man/systemd-sysext.html etc.

Reply:
In rpm-ostree's treefiles today to be nice we allow multiple packages in single YAML list entry.  See for example https://github.com/coreos/fedora-coreos-config/blob/aa50439847f34282b41bf2eb60d478a45d2a27d7/manifests/system-configuration.yaml#L7

Might make sense to do that too to avoid the overhead in the simple cases?

Reply:
We generally don't do that in _Ignition_ configs because it wouldn't work with the config merging logic, which (with one unfortunate exception) doesn't know the semantics of what it's merging.  That argument doesn't directly apply here, since this is sugar, but it still doesn't feel very Butane-like.  We've favored explicitness over ergonomics when they're in conflict.

Reply:
Followup in #316.